### PR TITLE
Convert tagPos to a size for address size computation

### LINKED ISSUE #14863: Tag position used instead of tag size to compute size of addresses
## Description

In [Assembly.cpp:L865-866](https://github.com/ethereum/solidity/blob/5fe31737472cc593cc9957d24a964ebf5491cbeb/libevmasm/Assembly.cpp#L865-L866), a tag position is compared with a tag size when the assembled object has subs. I.e., `tagPos` (a position) and `subTagSize` (a size) are compared and the maximum between those two values is kept. As a result, `bytesPerTag` and `bytesPerDataRef` sizes are miscomputed when the assembled object has some subs.

The effect is that sometimes, addresses that could fit in `PUSH1 XX` are pushed with `PUSH2 00YY`.

The bug is present with both via-ir and non via-ir modes.

## Environment

- Compiler version: 0.8.25
- Target EVM version (as per compiler settings): Shanghai
- Operating system: MacOS Sonoma

## Steps to Reproduce

Command: `./solc/solc --optimize --optimize-runs 200 --opcodes --bin ../contracts/Example.sol`

```solidity
pragma solidity <=0.8.25;

contract Example { 
    function b(uint256 x) external view returns(uint256) {
        uint256 y = block.timestamp;
        unchecked{ return y + x + gasleft(); }
    }
}
```

The init code generated for this contract is `6080604052348015600e575f80fd5b5060ca8061001b5f395ff3fe`, where `61001b` could be optimized into `601a`

### PR DESCRIPTION
**Pull request by simon-perriard**
The bug was fixed by converting the `tagPos` into its size.

Breakdown of changes:
-  `tagPos` is replaced by `numberEncodingSize(tagPos)` to determine the value of `subTagSize` in `Assembly::assemble()`

This fixes #14863

### CODE REVIEW

**Comment by nikola-matic** [comment]
Hi @simon-perriard. I'll be closing this PR since the issue you opened has not been triaged yet, and we have thus not yet decided whether this is a bug (most likely not), or just a missed optimization opportunity. Please keep track of the issue, and once we reach a decision, you can pick it up and open a new PR if you'd like.

**Comment by ekpyron** [comment]
The issue does indeed look genuine to me. We should add a Changelog entry to this - but then we can merge it!

It'd be in the ``Bugfix`` category, and the entry could be something like `` * Assembler: Prevent incorrect calculation of tag sizes.``.

Thank you very much for discovering this and for the fix!

**Comment by ekpyron** [comment]
I had a closer look also at the CI failures - most of them were just test expectation updates due to changing gas cost - but in fact the logic calculating the tag sizes didn't properly account for potentially referring to tags in subassemblies and that caused some cases to fail after this fix - so I took the liberty to push another fix for this together with test changes and a changelog entry on top of your initial fix.

File: libevmasm/Assembly.cpp [hunk 1]
```diff
@@ -891,6 +891,18 @@ LinkerObject const& Assembly::assemble() const
...
+		if (i.type() == PushTag)
+		{
+			auto [subId, tagId] = i.splitForeignPushTag();
+			if (subId == std::numeric_limits<size_t>::max())
+				continue;
+			assertThrow(subId < m_subs.size(), AssemblyException, "Invalid sub id");
+			for (auto subTagPosition: m_subs[subId]->m_tagPositionsInBytecode)
+				if (subTagPosition != std::numeric_limits<size_t>::max())
+					bytesPerTag = std::max(bytesPerTag, numberEncodingSize(subTagPosition));
+		}
```
Comment (ekpyron) [inline libevmasm/Assembly.cpp]: For the record: we need to adjust ``bytesPerTag`` to not only be sufficiently large to hold any tag into the current sources - but we also need to choose it large enough for tags to subassemblies (we use the same size for all tags - we could do something more clever in general, but that'd be a separate improvement).

Without this, some cases that actually contained references to subassembly tags of size 2 from an assembly with tags of size 1 in fact failed.

**Review by ekpyron — DISMISSED** [review]

**Review by ekpyron — APPROVED** [review]

**Pull request merged**